### ISSUE TITLE
Seeking Help to Resolve "endpoints_resolution_error" in @azure/msal-node@1.18.0

### ISSUE THREAD

**Issue by ghost** [AzureAD/microsoft-authentication-library-for-js#6239]
### Core Library

MSAL Node (@azure/msal-node)

### Core Library Version

1.18.0

### Wrapper Library

Not Applicable

### Wrapper Library Version

1.18.0

### Public or Confidential Client?

Confidential

### Description

Hello forum!

I have been struggling with an error for several days while attempting to resolve authentication endpoints in the @azure/msal-node@1.18.0 version. The specific error I'm encountering is "endpoints_resolution_error". I have meticulously checked the details of my application registered in Azure AD, including the client secret, Directory Identifier provided by the app, and the permissions granted to the API.

Despite my efforts, I haven't been able to resolve this issue, and I feel stuck. The code of ms-identity-javascript-nodejs-console works and has been what I have used to constrcut my code . 

I am reaching out to this community for help and guidance.  I would greatly appreciate any advice or solutions you can offer. 

Best regards,
lplazav

### MSAL Configuration

```javascript
async getToken() {

        console.log('Get Token from Microsoft...');
        console.log('- clientId: ' + this.CLIENT_ID);
        console.log('- authority: ' + this.AAD_ENDPOINT + this.TENANT_ID);
        console.log('- clientSecret : ' + this.CLIENT_SECRET);


        const tokenRequest = {
            scopes: [this.GRAPH_ENDPOINT + '.default'],
        };

        const msalConfig = {
            auth: {
                clientId: this.CLIENT_ID,
                authority: this.AAD_ENDPOINT + this.TENANT_ID,
                clientSecret: this.CLIENT_SECRET,
                knownAuthorities: [`${this.AAD_ENDPOINT}${this.TENANT_ID}`],
            },

            cache: {
                cacheLocation: "sessionStorage",
                storeAuthStateInCookie: false,
                secureCookies: false
            },
            system: {
                loggerOptions: {
                    logLevel: msal.LogLevel.Verbose,
                    loggerCallback: (level, message, containsPii) => {
                        if (containsPii) {
                            return;
                        }
                        switch (level) {
                            case msal.LogLevel.Error:
                                console.error(message);
                                return;
                            case msal.LogLevel.Info:
                                console.info(message);
                                return;
                            case msal.LogLevel.Verbose:
                                console.debug(message);
                                return;
                            case msal.LogLevel.Warning:
                                console.warn(message);
                                return;
                        }
                    },
                    piiLoggingEnabled: true
                },
            },
        };
        console.log(JSON.stringify(msalConfig, null, 2))
        const cca = new msal.ConfidentialClientApplication(msalConfig);
        console.log(JSON.stringify(cca, null, 2))
        return await cca.acquireTokenByClientCredential(tokenRequest).catch((err) => {
            throw err;
        });
    }
```

### Relevant Code Snippets

```javascript
INFO    11:15:14.662  811692cc33fec7bd  [Sat, 15 Jul 2023 11:15:14 GMT] : @azure/msal-node@1.18.0 : Info - acquireTokenByClientCredential called
DEBUG   11:15:14.662  811692cc33fec7bd  [Sat, 15 Jul 2023 11:15:14 GMT] : @azure/msal-node@1.18.0 : Verbose - initializeRequestScopes called
DEBUG   11:15:14.664  811692cc33fec7bd  [Sat, 15 Jul 2023 11:15:14 GMT] : [ae7ddda2-14cf-46a7-8150-84a5b8e19a9c] : @azure/msal-node@1.18.0 : Verbose - buildOauthClientConfiguration called
DEBUG   11:15:14.664  811692cc33fec7bd  [Sat, 15 Jul 2023 11:15:14 GMT] : [ae7ddda2-14cf-46a7-8150-84a5b8e19a9c] : @azure/msal-node@1.18.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/d1b91829-fc9d-4c6f-88f1-cdc17208686e
DEBUG   11:15:14.664  811692cc33fec7bd  [Sat, 15 Jul 2023 11:15:14 GMT] : [ae7ddda2-14cf-46a7-8150-84a5b8e19a9c] : @azure/msal-node@1.18.0 : Verbose - createAuthority called
DEBUG   11:15:14.666  811692cc33fec7bd  [Sat, 15 Jul 2023 11:15:14 GMT] : @azure/msal-node@1.18.0 : Verbose - Attempting to get cloud discovery metadata in the config
DEBUG   11:15:14.667  811692cc33fec7bd  [Sat, 15 Jul 2023 11:15:14 GMT] : @azure/msal-node@1.18.0 : Verbose - The host is included in knownAuthorities. Creating new cloud discovery metadata from the host.
DEBUG   11:15:14.667  811692cc33fec7bd  [Sat, 15 Jul 2023 11:15:14 GMT] : @azure/msal-node@1.18.0 : Verbose - Found cloud discovery metadata in the config.
INFO    11:15:19.718  811692cc33fec7bd  {
  "errorCode": "endpoints_resolution_error",
  "errorMessage": "Error: could not resolve endpoints. Please check network and try again. Detail: ClientAuthError: openid_config_error: Could not retrieve endpoints. Check your authority and verify the .well-known/openid-configuration endpoint returns the required endpoints. Attempted to retrieve endpoints from: https://login.microsoftonline.com/d1b91829-fc9d-4c6f-88f1-cdc17208686e/v2.0/.well-known/openid-configuration",
  "subError": "",
  "name": "ClientAuthError",
  "correlationId": "ae7ddda2-14cf-46a7-8150-84a5b8e19a9c"
}
```

### Identity Provider

Azure AD / MSA

### Source

External (Customer)

**Comment by bgavrilMS** [comment]
Hi @lplazav - the error indicates that MSAL JS is having trouble accessing the Internet. The library is trying to access: 

https://login.microsoftonline.com/d1b91829-fc9d-4c6f-88f1-cdc17208686e/v2.0/.well-known/openid-configuration

to read token_endpoint value from there. Unfortunately the logs do not give more details about the actual failure.

This type of problem happens a lot in environments with proxies. Is this your case?

Let's try the following: 

- change the network stack to use Axios. You do it like [this](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/msal-node-samples/custom-INetworkModule-and-network-tracing/app.ts#L38) and the Axios HTTPClient is [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/msal-node-samples/custom-INetworkModule-and-network-tracing/HttpClientAxios.ts)

**Comment by ghost** [comment]
Hi @bgavrilMS 

thank you very much for your time to answer me :)

I am not behind a proxy but behind a special runtime ( https://developer.atlassian.com/platform/forge/security/#sandboxing )

I have allowed by configuration to perform request to : 
`permissions:
  scopes:
    - storage:app
  external:
    fetch:
      backend:
        - 'https://api.atlassian.com'
        - 'https://login.microsoftonline.com'
        - 'https://graph.microsoft.com'
        - `

Testing with node-fetch, I have success when I do http GET to the url using this :

```let HTTP_STATUS, HTTP_STATUSTEXT;
  const myURL = "https://login.microsoftonline.com/4b876178-13a5-4e1b-82c0-a1b8a0c390e7/v2.0/.well-known/openid-configuration";

  fetch(
    myURL,
    {
      headers: {
        "Content-Type": "application/json",
      },
    },
  )
    .then((res) => {
      HTTP_STATUS = res.status;
      HTTP_STATUSTEXT = res.statusText;
      return res.json();
    })
    .then((HTTP_BODY) => {
      if (HTTP_STATUS == 200) {
        console.log(HTTP_BODY)

        resolve(true);
      } else {
        reject(`An error has been detected: ${HTTP_STATUS} - ${HTTP_STATUSTEXT}`);
      }
    })
    .catch((err) => {
      // handle error
      reject(err);
    });
```

I have these references in my sources:

```
  "dependencies": {
    "@azure/msal-node": "^1.8.0",
    "@forge/api": "^2.18.1",
    "@forge/ui": "^0.13.1"
  },
```
`import * as msal from "@azure/msal-node";
`
...
but get:
`ClientAuthError: endpoints_resolution_error: Error: could not resolve endpoints. Please check network and try again. Detail: ClientAuthError: openid_config_error: Could not retrieve endpoints. Check your authority and verify the .well-known/openid-configuration endpoint returns the required endpoints. Attempted to retrieve endpoints from: https://login.microsoftonline.com/4b876178-13a5-4e1b-82c0-a1b8a0c390e7/v2.0/.well-known/openid-configuration`

Do you think is something I miss? The "issuer" url value is not reachable "issuer": "https://login.microsoftonline.com/4b876178-13a5-4e1b-82c0-a1b8a0c390e7/v2.0",

**Comment by bgavrilMS** [comment]
Can you try to use the Axios HttpClient and see if all the calls work as you expect?

**Comment by bgavrilMS** [comment]
@Robbie-Microsoft - any suggestions for this issue? Do you have enough log data to understand what's happening?

**Comment by Robbie-Microsoft** [comment]
I don't have any suggestions other than what you said @bgavrilMS. The logs aren't clear. @lplazav, did you try Bogdan's suggestion and use the Axios HttpClient and see if the network calls work as expected? You can also implement your own network module.

**Comment by bgavrilMS** [comment]
Can we get more data into those logs @Robbie-Microsoft ?

**Comment by ghost** [comment]
RESOLVED, (    "@azure/msal-node": "^2.0.1", ) now I get more log details and I can get the token and fetch the data, at last it was not needed to implement HttpClient ( I am using node-fetch not Axios ). Very grateful for your help!

INFO    23:20:49.817  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Info - acquireTokenByClientCredential called
DEBUG   23:20:49.817  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Verbose - initializeRequestScopes called
DEBUG   23:20:49.819  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : [9d41f476-aa4b-4011-ade6-ae004b46b694] : @azure/msal-node@2.0.1 : Verbose - buildOauthClientConfiguration called
DEBUG   23:20:49.819  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : [9d41f476-aa4b-4011-ade6-ae004b46b694] : @azure/msal-node@2.0.1 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/d1b91829-fc9d-4c6f-88f1-cdc17208686e
DEBUG   23:20:49.819  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : [9d41f476-aa4b-4011-ade6-ae004b46b694] : @azure/msal-node@2.0.1 : Verbose - createAuthority called
DEBUG   23:20:49.820  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Verbose - Attempting to get cloud discovery metadata  from authority configuration
DEBUG   23:20:49.821  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Verbose - The host is included in knownAuthorities. Creating new cloud discovery metadata from the host.
DEBUG   23:20:49.821  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Verbose - Found cloud discovery metadata in authority configuration
DEBUG   23:20:49.821  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Verbose - Attempting to get endpoint metadata from authority configuration
DEBUG   23:20:49.821  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Verbose - Did not find endpoint metadata in the config... Attempting to get endpoint metadata from the hardcoded values.
DEBUG   23:20:49.822  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Verbose - Did not find endpoint metadata in hardcoded values... Attempting to get endpoint metadata from the network metadata cache.
DEBUG   23:20:49.822  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Verbose - Did not find cached endpoint metadata... Attempting to get endpoint metadata from the network.
DEBUG   23:20:49.822  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:49 GMT] : @azure/msal-node@2.0.1 : Verbose - Authority.getEndpointMetadataFromNetwork: attempting to retrieve OAuth endpoints from https://login.microsoftonline.com/d1b91829-fc9d-4c6f-88f1-cdc17208686e/v2.0/.well-known/openid-configuration
DEBUG   23:20:50.740  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:50 GMT] : @azure/msal-node@2.0.1 : Verbose - Endpoint metadata was successfully returned from getEndpointMetadataFromNetwork()
DEBUG   23:20:50.741  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:50 GMT] : [9d41f476-aa4b-4011-ade6-ae004b46b694] : @azure/msal-node@2.0.1 : Verbose - Client credential client created
DEBUG   23:20:51.684  d8d6998b-7824-4457-a06f-49c17124c5c7  [Sat, 12 Aug 2023 21:20:51 GMT] : [9d41f476-aa4b-4011-ade6-ae004b46b694] : @azure/msal-common@14.0.1 : Verbose - No client info in response
INFO    23:20:51.686  d8d6998b-7824-4457-a06f-49c17124c5c7  Get users from Azure
INFO    23:20:51.687  d8d6998b-7824-4457-a06f-49c17124c5c7  get users from ...$top=999&$select=displayName,jobTitle,department,officeLocation,companyName,streetAddress,postalCode,city,state,country,mail,telephoneNumber,businessPhones,mobile,objectId,id,accountEnabled,deletionTimestamp,dirSyncEnabled,givenName,surname,isCompromised,lastDirSyncTime,mailNickname,objectType,preferredLanguage,usageLocation,userPrincipalName,userState,userType,manager
INFO    23:20:51.687  d8d6998b-7824-4457-a06f-49c17124c5c7  fetching data...
